vllm/model_executor/layers/attention/ops/flash_attention_triton.py

```diff
@@ -0,0 +1,541 @@
+        # This is a > check because mask being 0 blocks the store.
+        l_ptrs_mask = boundary > tl.arange(0, BLOCK_M)
+        tl.store(l_ptrs, m_i + tl.math.log2(l_i), mask=l_ptrs_mask)
+    else:
+        tl.store(l_ptrs, m_i + tl.math.log2(l_i))
+
+    # write back O
+    o_offset = off_z * stride_oz + cu_seqlens_q_start * stride_om + off_h_q * stride_oh
+    O_block_ptr = tl.make_block_ptr(
+        base=Out + o_offset,
+        shape=(seqlen_q, ACTUAL_BLOCK_DMODEL),
+        strides=(stride_om, stride_on),
+        offsets=(start_m * BLOCK_M, 0),
+        block_shape=(BLOCK_M, BLOCK_DMODEL),
+        order=(1, 0)
+    )
+    # Need boundary check on this to make sure the padding from the 
+    # Q and KV tensors in both dims are not part of what we store back.
+    # TODO: Do the boundary check optionally. 
+    tl.store(O_block_ptr, acc, boundary_check=(0,1))
+
+def check_args(q, k, v, o, max_seqlens):
+        assert q.dim() == k.dim() and q.dim() == v.dim()
+        assert q.dim() == 4
+        batch, nheads_q, seqlen_q, head_size = q.shape
+        _, nheads_k, seqlen_k, _ = k.shape
+        assert max_seqlens > 0
+        assert k.shape == v.shape
+        assert q.shape[-1] == k.shape[-1] and q.shape[-1] == v.shape[-1]
+        # TODO: Change assert if we support qkl f8 and v f16
+        assert q.dtype == k.dtype and q.dtype == v.dtype
+        # TODO: Fix assert to check head size <=256 once supported
+        assert head_size <= 128
+        assert o.shape == q.shape
+        assert (nheads_q % nheads_k) == 0
+
+class _attention(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, q, k, v, o, metadata, causal=False, sm_scale=1.0, bias=None):
+        if o is None:
+            o = torch.empty_like(q, dtype=v.dtype)
+        check_args(q, k, v, o, metadata.max_seq_len)
+
+        batch, seqlen_q, nheads_q, head_size = q.shape
+        _, seqlen_k, nheads_k, _ = k.shape
+        q_strides = (q.stride(0), q.stride(2), q.stride(1), q.stride(3))
+        k_strides = (k.stride(0), k.stride(2), k.stride(1), k.stride(3))
+        v_strides = (v.stride(0), v.stride(2), v.stride(1), v.stride(3))
+        o_strides = (o.stride(0), o.stride(2), o.stride(1), o.stride(3))
+
+        # Get closest power of 2 over or equal to 32.
+        unpadded_head_dims = {32, 64, 128}
+        if head_size not in unpadded_head_dims:
+            padded_d_model = None
+            for i in unpadded_head_dims:
+                if i > head_size:
+                    padded_d_model = i
+                    break
+            assert padded_d_model is not None
+        else:
+            padded_d_model = head_size
+
+
+        grid = lambda META: (
+            triton.cdiv(metadata.max_seq_len, META['BLOCK_M']),
+            nheads_q,
+            batch
+        )
+
+        encoded_softmax = None
+
+        M = torch.empty((batch, nheads_q, metadata.max_seq_len), device=q.device, dtype=torch.float32)
```

Comment:
I think you can remove this, and the associated writeback in the kernel for a minor speedup. This is used during the backward pass and we don't have a separate inference-only forward kernel. We don't need it for inference.